### DISCUSSION TITLE
SMW Physics/Mechanics

### DISCUSSION THREAD

**Discussion by 0lhi** [Wohlstand/TheXTech#157] [Ideas]
The Physics of SMBX are generally deemed inferior to those of SMW. I've seen it brought up in Streams, the [SMBX Forums](https://www.supermariobrosx.org/forums/viewtopic.php?f=69&t=24717), the [Wohlsoft Forums](https://wohlsoft.ru/forum/viewtopic.php?f=11&t=71&p=4636#p4619), and I experienced myself [more than once](https://wohlsoft.ru/pgewiki/User:Olhi) that enjoyment of an Episode was decreased or prevented due to Mario's slow acceleration.

In the Wohlsoft Post from 2016, sken130 made a list of suggestions, some of which I would like to pass on here:

> 1. In SMBX, while running, Mario's speed is accelerated constantly towards the maximum. While in SMW, Mario's horizontal speed accelerates from 0 to 36, and stay in 36 for 1-2 seconds, and jumps (not accelerates smoothly) from 36 to 48 finally. Remarks: Mario's normal walking speed on land is 20.
> 2. This's for swimming physics. In SMBX, pressing the jump button has the same effect as pressing the jump+up button. But in SMW, pressing jump+up button will swim upwards much faster than pressing only the jump button, and pressing jump+down button will swim only slightly upwards. Also, in SMW, while carrying items (Koopa shells, P-Switches, Spring-boxes, etc.), Mario's swimming horizontal speed increases from 16 to 32. Mario's walking speed under water is 8.
> 9. Mario's acceleration is higher SMW than in SMBX, making Mario more agile in SMW than in SMBX. I prefer "difficulty due to strong enemies" (which is more interesting since Mario rocks in this way) rather than "difficulty due to slow Mario" (which is a fake difficulty and Mario sucks in this way).
> 
> For reference, you can feel Mario's agility in the extremely hard final boss fight of Kaizo Mario 3 (https://www.youtube.com/watch?v=oYHlznTM2zo), where the Japanese player R.Kiba lost around 110 lives there, and at 7:30 Mario's agility is fully shown. As a comparison, one would expect the final boss of SMBX's Princess Cliche 2 would be much easier (even without the bullet bill gun nor the hammer suit) because in Princess Cliche 2 Bowser doesn't throw 5 hammers, doesn't spit 3 fire balls, doesn't jump quickly, and doesn't have Magikoopa and Thwomp as allies. But still it's not easy to fight, and I lost more than 150 lives in the final boss of Princess Cliche 2 simply because Mario in SMBX doesn't have enough acceleration and agility (even if Mario has enough max speed) to dodge the flying koopas. Mario also has a bigger hit box in SMBX than in SMW. To summarize, SMBX physics won't allow agile Mario movements similar to Kaizo 3 boss fight 7:30.
> 
> With SMW physics, the final boss of SMBX's Princess Cliche 2 would be a piece of cake, and we can make it harder by putting more enemies or obstacles.

The SMBX2 Campaign [A Tiny World Episode](https://www.supermariobrosx.org/forums/viewtopic.php?f=36&t=27087) includes a Physics Patch that is described as "SMW Physics" on the [Speedrun Leaderboards](https://www.speedrun.com/smbx_atwe). I don't know how close it actually is to the SNES game, but it offers an example for how more precise controls could look like:

![2021-07-16_17_58_01](https://user-images.githubusercontent.com/74732674/125978670-dd73a139-bb0a-458e-9b2a-f631d9923bfe.gif)

Mario is the only allowed Character, but it seems the Patch affects everyone:

Mario: 4.6 / 4.7

Luigi: 4.7 / 4.8

Toad: 4.3 / 4.4

Peach: 5.007 / 5.085

Link: 4.9 / 5.0

If it'd be like that for TheXTech, too, I don't think it would be a bad thing.

Like in the episode, the player should be able to switch at any time. With that being the case, I don't think a compat.ini entry is needed, since neither mechanics shouldn't be forced on the player, unless in Mode 2 or 3, which should default to SMBX-Physics.

(Mode 2 may be debatable since theoretically, an episode could aim for cross-compatibility while providing a Physics Patch for X2. In that case, a compat.ini that allows to choose even in Speedrunning Mode could be added. But, that scenario doesn't have practical significance now.)

Also, in [this thread about 1.3 Physics](https://wohlsoft.ru/forum/viewtopic.php?f=64&t=3833), Lx Xzit brought up this point on Toad:

> Toad is a weird case because when he reaches his maximum walking speed (for short, MWS), his speed decreases to 3.12 and then increases to 3.22 again. So we can conclude his MWS value oscillates between 3.12 and 3.22. However, I think this shouldn't happen in SMBX 1.4, it would be better if his MSW value is just set to 3.22. There's no reason to do exactly what Redigit did. 

Maybe this is something to take a look at as well?

**Comment by ChristianSilvermoon** [comment]
I actually _strongly_ prefer SMBX's physics to SMW's. For me, they feel (not claiming accuracy or that they are) somewhere between SMB3 and SMW in a sort of comfortable sweet spot. If they were changed, I'd want it to be optional.

Also, the swimming controls with up and down are a thing in SMBX :P
![Scr_2021-08-23_15-34-29](https://user-images.githubusercontent.com/30162808/130528159-1212d7a7-17c6-4797-ab54-d96a69577832.gif)
EDIT: I didn't notice you'd mentioned optional somehow, I'm a derp XD

**Reply by 0lhi** [reply]
> Also, the swimming controls with up and down are a thing in SMBX :P

Huh, good to know :o

> EDIT: I didn't notice you'd mentioned optional somehow, I'm a derp XD

:P

**Comment by ChristianSilvermoon** [comment]
Alternative suggestion:
a `physics.ini` allowing creators to customize per player character physics/rules

IE top walk speed, run speed, friction, jump height, acceleration, how long before flight can be done, pick up speed, gravity, etc.

Could be level and/or episode specific.

This adds the possibility of SMW-like physics as well as many other customized physics options, such as allowing low or high gravity levels

**Reply by Wohlstand** [reply]
Seems a good idea, at least, all physics related arrays were available here and can be tuned. It's clearly understandible and doesn't looks like the suspicious hidden ASM-coded structure like we had while worked on LunaLua.

**Reply by Wohlstand** [reply]
Additionally, I want to add support for extra SMBX-38A environments that can be very useful for custom liquids, on-enter event triggers, etc. (there is "Air" environment even exist that can allow making chambers at fully underwater sections, with less effort)

**Reply by ChristianSilvermoon** [reply]
I think the configuration changes should be cumulative as well, so you only need to specify what you want to change in the ini file, with changes from the 'most recently loaded' taking priority, like so:

1. Level Specific physics
2. Episode Specific physics
3. Game-wide base physics

So if you customize physics for a whole episode, but want one low gravity level, you could just change the gravity in that level... though perhaps level specific physics should/could also (or exclusively) be handled in level-file and set within the editor? :thinking:

**Reply by 0lhi** [reply]
If this is the approach, there should also be a way to invert controls. That is a feature of Super Mario Enigmatic 2, an amazing X2 Episode I want to port to TheXTech someday.

Physics Settings should be also changeable by Events.

**Comment by 0lhi** [comment]
This idea no longer aligns with the priorities of TheXTech's development.

Similar and alternative ideas may be discussed here: #636
